src/GrafanaDashboardDeploy/src/Util.cs

```diff
@@ -0,0 +1,103 @@
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace DotNet.Grafana
+{
+    /// <summary>
+    /// Utility class to hold methods manipulating JSON in ways specific to the Grafana API
+    /// </summary>
+    public class Util
+    {
+        /// <summary>
+        /// Extract the Folder ID of a Dashboard from a JSON object returned by the api/dashboards/uid endpoint
+        /// </summary>
+        /// <param name="dashboard"></param>
+        /// <returns></returns>
+        public static int ExtractFolderId(JObject dashboard)
+        {
+            return dashboard["meta"]["folderId"].Value<int>();
+        }
+
+        /// <summary>
+        /// Modify a Dashboard JSON object as retrieved from the Grafana API into
+        /// something suitable to post back to the API
+        /// </summary>
+        public static JObject SanitizeDashboard(JObject dashboard)
+        {
+            var slimmedDashboard = new JObject(dashboard["dashboard"].ToObject<JObject>());
+            slimmedDashboard.Remove("id");
+            slimmedDashboard.Remove("version");
+            return slimmedDashboard;
+        }
+
+        /// <summary>
+        /// Extract the names of data sources used by a given dashboard. 
+        /// </summary>
+        /// <param name="dashboard">A JSON definition of a dashboard as delivered by the Grafana API</param>
+        /// <returns></returns>
+        public static IEnumerable<string> ExtractDataSourceNames(JObject dashboard)
+        {
+            // Datasources live in panel[*].datasource, unless the "Mixed Datasource" feature
+            // is used. Then, get names from panel[*].target.datasource. 
+            var datasourceNames = dashboard
+                .SelectTokens("$.dashboard.panels[*]..datasource")
+                .Values<string>()
+                .Where(x => !String.IsNullOrEmpty(x))
+                .Where(x => x != "-- Mixed --")
+                .Distinct();
+
+            return datasourceNames;
+        }
+        
+        /// <summary>
+        /// Modify a Data Source JSON object as retrieved from the Grafana API into
+        /// something suitable to post back to the API
+        /// </summary>
+        public static JObject SanitizeDataSource(JObject dataSource)
+        {
+            JObject slimmedDatasource = new JObject(dataSource);
+            slimmedDatasource.Remove("id");
+            slimmedDatasource.Remove("orgId");
+            slimmedDatasource.Remove("url");
+
+            // Add an entry in secureJsonData for each secureJsonField and decorate as a KeyVault insert.
+            if (dataSource.ContainsKey("secureJsonFields"))
+            {
+                var secureJsonData = new JObject();
+                foreach (var secretField in dataSource["secureJsonFields"].Children<JProperty>())
+                {
+                    secureJsonData.Add(new JProperty(secretField.Name, $"[vault({secretField.Name})]"));
```

Comment:
Isn't the secret field just going to be something horribly generic, like "apiKey"? That's not a very good secret name to put i the vault (especially since there will be lots of "apiKey" values)

Reply:
Oooooh. The person has to fill them in.  Makes sense... Does this only generate the placeholders for "new" datasources (and existing ones are unaltered)